aspnetcore/blazor/forms-validation.md

```diff
@@ -223,12 +230,383 @@ In the following example:
+        {
+            errors.Add(nameof(starship.Description),
+                new List<string>() { "For a 'Defense' ship classification, " +
+                "'Description' is required." });
+        }
+
+        if (errors.Count() > 0)
+        {
+            businessValidator.DisplayErrors(errors);
+        }
+        else
+        {
+            // Process the form
+        }
+
+        formStatus = $"Form submit: {DateTime.Now}";
+    }
+}
+```
+
+## Server validation
+
+Server validation can be accomplished with a custom server validation component:
+
+* Send the <xref:Microsoft.AspNetCore.Components.Forms.EditContext.Model?displayProperty=nameWithType> to a backend server API.
+* Process model validation on the server when the form passes client-side validation (<xref:Microsoft.AspNetCore.Components.Forms.EditForm.OnValidSubmit>).
+* Return field validation errors to the client app in a dictionary.
+* Display the errors in the form with a custom validation component.
+
+The following example is based on:
+
+* A hosted Blazor solution created by the Blazor Hosted project template.
+* The *Starfleet Starship Database* form example in the preceding [Built-in forms components](#built-in-forms-components) section.
+
+In the following example:
+
+* The server API validates that the value of the ship description (`Description`) isn't `null` or empty if the user selects the `Defense` ship classification (`Classification`).
+* The placeholder `{SOLUTION NAMESPACE}` is the solution's namespace.
+
+Place the `Starship` model into the solution's `Shared` project so that both the client and server apps can use the model. Since the model requires data annotations, add a package reference for [`System.ComponentModel.Annotations`](https://www.nuget.org/packages/System.ComponentModel.Annotations) to the `Shared` project's project file:
+
+```xml
+<ItemGroup>
+  <PackageReference Include="System.ComponentModel.Annotations" Version="{VERSION}" />
+</ItemGroup>
+```
+
+Replace the `{VERSION}` placeholder with the correct version of the package.
+
+In the server API project, add a controller to process starship validation requests (`Controllers/StarshipValidation.cs`) and return failed validation messages:
+
+```csharp
+using System;
+using System.Collections.Generic;
+using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.Extensions.Logging;
+using {SOLUTION NAMESPACE}.Shared;
+
+namespace {SOLUTION NAMESPACE}.Server.Controllers
+{
+    [Authorize]
+    [ApiController]
+    [Route("[controller]")]
+    public class StarshipValidationController : ControllerBase
+    {
+        private readonly ILogger<StarshipValidationController> logger;
+
+        public StarshipValidationController(
+            ILogger<StarshipValidationController> logger)
+        {
+            this.logger = logger;
+        }
+
+        [HttpPost]
+        public Dictionary<string, List<string>> Post(Starship starship)
```

Comment:
you should use ModelState so it would be more integrated with the framework tools